File: ruby/shell.rb
class Environment

def initialize()
  @globals = {:h_limit=>10}
  @prompt = "ruby  shell> "
  @hist = []
end

def run(exp)
  unless exp.to_s.strip.empty?
    eval exp
    @hist.push exp
  end
end

def addExp(exp)
  @hist.push exp
end

def historyIndexFromList(offset)
  index = @hist.length - 1 - offset
  index < 0 ? @hist.length : index
end

def edit(offset, pat, sub)
  cmd = ''
  if @hist.length == 0
    print "Empty history\n"
  elsif not pat
    print "No pattern regex provided\n";
  elsif not sub
    print "No substitution regex provided\n";
  else
    index = historyIndexFromList (offset ? offset - 1 : 0)
    cmd = @hist[index]
    cmd.sub(pat,sub)
    puts cmd
  end
  cmd
end

def showHelp
  print "help - this list\n";
  print "history {n} - n lines of history (empty == all)\n";
  print "h - history #{@globals[:h_limit]}\n";
  print "!{n} - repeat nth previous command (! == !1)\n";
  print "!{n}e s/pat/sub/ - edit and run nth previous command using regexp\n";
end

def showHistory(limit)
  limit = limit == -1 ? @globals[:h_limit] : limit
  start = limit ? @hist.length - limit : 0
  start = start < 0 ? 0 : start
  it = start
  while it < @hist.length
    puts "#{@hist.length - it}  #{@hist[it]}"
    it+=1
  end
end

def history(offset)
  index = historyIndexFromList offset
  cmd = @hist[index]
  if @hist.length == 0
    puts "History empty"
  elsif cmd
    puts cmd
  else
    puts "Bad command: ! #{offset + 1}"
  end
  cmd
end

end #Environment


def loadFile(file)
  load file
  print "Loaded #{file}"
end

prompt = "ruby shell> "
e = Environment.new
print prompt

while exp = gets.chomp
  case exp
  when /^\s*history\s*$/
    e.showHistory 0
  when 'h'
    e.showHistory -1
  when /\s*history\s+(\d+)\s*$/
    e.showHistory $1.to_i
  when /^\s*!(\d*)e\s*s\/(.*)\/(.*)\/\s*$/
    exp = e.edit($1.to_i, $2, $3)
    e.run exp
  when '!'
    exp = e.history 0
    e.run exp
  when /!(\d+)\s*$/
    exp = e.history($1.to_i - 1)
    e.run exp
  when /^\s*help\s*/
    e.showHelp
  when /\s*load\s*\(\s*['"](\w+)['"]\s*\)/
    loadFile($1)
    e.addExp exp
  when 'quit'
    break
  when 'exit'
    break
  else
    e.run exp
  end
  print prompt
end

File: python/shell.py
import re

GLOBALS = {'h_limit': 10}
regex_load = r'\s*load\s+(\S+)\s*'
regex_repeat_digit = r'!(\d+)\s*$'
regex_edit = r'^\s*!(\d*)e\s*s\/(.*)\/(.*)\/\s*$'
regex_history_digit = r'\s*history\s+(\d+)\s*$'

prompt = "\npython shell> "

exp = ''
history = []


def run(exp):
    if exp:
        eval(exp)
        history.append(exp)


def historyIndexFromList(offset):
    index = len(history) - 1 - offset
    return len(history) if index < 0 else index


def edit(offset, pat, sub):
    cmd = ''
    if len(history) == 0:
        print("Empty history\n")
    elif not pat:
        print("No pattern regex provided\n")
    elif not sub:
        print("No substitution regex provided\n")
    else:
        index = historyIndexFromList((offset - 1) if offset else 0)
        cmd = history[index]
        cmd = re.sub(pat, sub, cmd)
        print(cmd)
    return cmd


def showHelp():
    print("help - this list")
    print("history {n} - n lines of history (empty == all)")
    print("h - history ", GLOBALS['h_limit'])
    print("!{n} - repeat nth previous command (! == !1)")
    print("!{n}e s/pat/sub/ - edit and run nth previous command by replacing pat with sub")


def showHistory(limit):
    start = (len(history) - limit) if limit else 0
    start = 0 if start < 0 else start
    it = start
    while it < len(history):
        print(len(history) - it, " ", history[it])
        it += 1


def execHistory(offset):
    index = historyIndexFromList(offset)
    cmd = history[index]
    if len(history) == 0:
        print("History empty")
    elif cmd:
        print(cmd)
    else:
        print("Bad command: !", offset + 1)

    return cmd


def load(file):
    f = open(file)
    source = f.read()
    exec(source)
    print("Loaded ", file)


# Execute loop
while True:
    exp = input(prompt).strip()
    if re.search(r'^\s*history\s*$', exp):
        showHistory()
    elif exp == 'h':
        showHistory(GLOBALS['h_limit'])
    elif re.search(regex_history_digit, exp):
        m = re.match(regex_history_digit, exp)
        showHistory(int(m.group(1)))
    elif re.search(regex_edit, exp):
        m = re.match(regex_edit, exp)
        exp = edit(int(m.group(1)), str(m.group(2)), str(m.group(3)))
        run(exp)
    elif exp == '!':
        exp = execHistory(0)
        run(exp)
    elif re.search(regex_repeat_digit, exp):
        m = re.match(regex_repeat_digit, exp)
        exp = execHistory(int(m.group(1)) - 1)
        run(exp)
    elif re.search(r'^\s*help\s*', exp):
        showHelp()
    elif re.search(regex_load, exp):
        m = re.match(regex_load, exp)
        load(m.group(1))
        history.append(exp)
    elif exp == 'quit' or exp == 'exit':
        break
    else:
        run(exp)
